src/CodeGen_Vulkan_Dev.cpp

```diff
@@ -0,0 +1,1304 @@
+    void add_kernel(Stmt stmt,
+                    const std::string &name,
+                    const std::vector<DeviceArgument> &args) override;
+
+    /** (Re)initialize the GPU kernel module. This is separate from compile,
+     * since a GPU device module will often have many kernels compiled into it
+     * for a single pipeline. */
+    void init_module() override;
+
+    std::vector<char> compile_to_src() override;
+
+    std::string get_current_kernel_name() override;
+
+    void dump() override;
+
+    std::string print_gpu_name(const std::string &name) override;
+
+    std::string api_unique_name() override {
+        return "vulkan";
+    }
+
+protected:
+    class SPIRVEmitter : public IRVisitor {
+
+    public:
+        SPIRVEmitter() = default;
+
+        using IRVisitor::visit;
+
+        void visit(const Variable *) override;
+        void visit(const IntImm *) override;
+        void visit(const UIntImm *) override;
+        void visit(const StringImm *) override;
+        void visit(const FloatImm *) override;
+        void visit(const Cast *) override;
+        void visit(const Add *) override;
+        void visit(const Sub *) override;
+        void visit(const Mul *) override;
+        void visit(const Div *) override;
+        void visit(const Mod *) override;
+        void visit(const Max *) override;
+        void visit(const Min *) override;
+        void visit(const EQ *) override;
+        void visit(const NE *) override;
+        void visit(const LT *) override;
+        void visit(const LE *) override;
+        void visit(const GT *) override;
+        void visit(const GE *) override;
+        void visit(const And *) override;
+        void visit(const Or *) override;
+        void visit(const Not *) override;
+        void visit(const Call *) override;
+        void visit(const Select *) override;
+        void visit(const Load *) override;
+        void visit(const Store *) override;
+        void visit(const Let *) override;
+        void visit(const LetStmt *) override;
+        void visit(const AssertStmt *) override;
+        void visit(const ProducerConsumer *) override;
+        void visit(const For *) override;
+        void visit(const Ramp *) override;
+        void visit(const Broadcast *) override;
+        void visit(const Provide *) override;
+        void visit(const Allocate *) override;
+        void visit(const Free *) override;
+        void visit(const Realize *) override;
+        void visit(const IfThenElse *) override;
+        void visit(const Evaluate *) override;
+        void visit(const Shuffle *) override;
+        void visit(const Prefetch *) override;
+        void visit(const Fork *) override;
+        void visit(const Acquire *) override;
+
+        void visit_binop(Type t, const Expr &a, const Expr &b, uint32_t opcode);
+
+        // ID of last generated Expr.
+        uint32_t id;
+        // IDs are allocated in numerical order of use.
+        uint32_t next_id{0};
```

Comment:
nit: we don't configure clang-tidy to enforce this, but I generally we prefer `next_id = 0;` style for in-class-member-init, rather than `next_id{0};`